### ISSUE TITLE
Validate API responses

### ISSUE THREAD

**Issue by ErikMichelson** [hedgedoc/react-client#1219]
**Describe the solution you'd like**
API response objects are type-safe at compile-time but the actual API responses don't get validated, invalid input could crash the app. We should therefore add runtime checks for the actual API response data.
To avoid generating schemata on our own, we could use https://github.com/ForbesLindesay/typescript-json-validator. We could generate the schemata via the CLI of that package on build-time.

**Describe alternatives you've considered**
Write the schemata on our own and feed them to e.g. joi. While this is more consistent with the way the backend does it, it could make trouble when the schema changes and the types were forgotten to update or the otherway around.

**Comment by davidmehren** [comment]
Is there maybe a way to leverage the OpenAPI Spec from the backend to automatically generate validation code?

**Comment by mrdrogdrog** [comment]
This is not what we want.
We want to validate that the response matches the typescript types that are defined in the frontend. Because then we can make sure that the build time type constraints are also valid on runtime.

If the types of frontend and backend are not the same (because we forgot to adjust the types in the frontend after a backend minor change) then the CI would fail and we know that there are some type mismatches that have to be fixed.

**Comment by mrdrogdrog** [comment]
E.g. you change a key in the frontend config response in the backend (like changing specialUrls to awesomeUrls). Then the frontend CI would fail because we expect "specialUrls" but it's not there. Otherwise the runtime would try to process a value that should not be undefined but is. => 💥

**Comment by davidmehren** [comment]
If you have
- an OpenAPI schema
- some kind of validator generated from that, 
- frontend code that uses the types of the validator

I think you get exactly what you want.

If the behavior of the backend changes without somebody updating the validator (using the OpenAPI schema), the validation will fail at runtime.
If the validator is updated and the types now stop matching the usage in the frontend, the frontend won't compile.

**Comment by davidmehren** [comment]
The frontend could also just use the `.dto.ts` files from the backend and use them to validate JSON with [class-validator](https://github.com/typestack/class-validator). That would skip the OpenAPI step and allow code-reuse if frontend and backend code would ever reside in the same repo.

**Comment by mrdrogdrog** [comment]
But because frontend and backend are built and developed separately this wouldn't change anything. There is still the manual interaction "generate new validators/types from the API spec". This means that our proposed method would still work and your proposal would be the additional step "create types from the openapi spec".
The only difference is that we would do in another order.
Openapi => types => validators .

**Comment by mrdrogdrog** [comment]
> The frontend could also just use the `.dto.ts` files from the backend and use them to validate JSON with [class-validator](https://github.com/typestack/class-validator). That would skip the OpenAPI step and allow code-reuse if frontend and backend code would ever reside in the same repo.

Yes but they aren't in the same repo. So we want to validate the responses.

**Comment by mrdrogdrog** [comment]
Okay for clarification:

We're both talking about runtime validation, but David proposed to copy the types from the backend so we don't have to create them again.

Well. That's also a possibility
